chemclipse/plugins/org.eclipse.chemclipse.xxd.model/src/org/eclipse/chemclipse/xxd/model/filter/peaks/ShapeFilter.java

```diff
@@ -0,0 +1,219 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Lablicate GmbH.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ * Alexander Stark - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.chemclipse.xxd.model.filter.peaks;
+
+import java.util.Collection;
+import java.util.function.BiPredicate;
+import org.eclipse.chemclipse.model.core.IPeak;
+import org.eclipse.chemclipse.model.core.IPeakModel;
+import org.eclipse.chemclipse.model.filter.IPeakFilter;
+import org.eclipse.chemclipse.processing.Processor;
+import org.eclipse.chemclipse.processing.core.MessageConsumer;
+import org.eclipse.chemclipse.processing.filter.CRUDListener;
+import org.eclipse.chemclipse.processing.filter.Filter;
+import org.eclipse.chemclipse.xxd.model.support.PeakShapeFilterSelectionCriterion;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.SubMonitor;
+import org.osgi.service.component.annotations.Component;
+
+import com.google.common.collect.Range;
+
+@Component(service = { IPeakFilter.class, Filter.class, Processor.class })
+public class ShapeFilter implements IPeakFilter<ShapeFilterSettings> {
+
+	private static BiPredicate<Double, Double> LEADING_SMALLER_THAN_LIMIT_COMPARATOR = (value, shapeSetting) -> (value < shapeSetting);
+	private static BiPredicate<Double, Double> TAILING_GREATER_THAN_LIMIT_COMPARATOR = (value, shapeSetting) -> (value > shapeSetting);
+	private static BiPredicate<RangeContainer, Range<Double>> VALUES_WITHIN_RANGE_COMPARATOR = (container, shapeSetting) -> (shapeSetting.contains(container.leading) && (shapeSetting.contains(container.tailing)));
+
+	private static class ShapePredicate<T> {
+
+		private final BiPredicate<Double, T> predicate;
+		private final T shapeSetting;
+
+		public ShapePredicate(BiPredicate<Double, T> predicate, T shapeSetting) {
+
+			super();
+			this.predicate = predicate;
+			this.shapeSetting = shapeSetting;
+		}
+
+		public ShapePredicate<?> negate() {
+
+			return new ShapePredicate<T>(predicate.negate(), shapeSetting);
+		}
+
+		public boolean test(double value) {
+
+			boolean result = predicate.test(value, shapeSetting);
+			return result;
+		}
+	}
+
+	private static class RangeContainer {
```

Comment:
This could be replaced by Guava's 'Range' class.

Reply:
Are you sure?

This is obviously an object to store two different peak property values that change for every inspected peak. We don't want to construct a range here because we are solely interested in these two values rather than a range.

Reply:
Moreover, as stated in the Guava Range javadoc: _A range (or "interval") defines the boundaries around a contiguous span of values of some Comparable type; for example, "integers from 1 to 100 inclusive."_

That means we cannot construct/initiate half of the needed ranges because the leading is greater than tailing. Therefore I will stick to the RangeContainer but also try to use one predicate as proposed in the next comment.

Reply:
![grafik](https://user-images.githubusercontent.com/39908870/73747509-dcc75800-4757-11ea-8048-4e8efe6803f6.png)

However, this also means that if you want to filter according to the values ​​in the Peak/Scan List, the settings initialization will also fail.

Reply:
correct, it is not a range.